kernel/src/log_replay.rs

```diff
@@ -190,3 +194,100 @@ impl<'seen> FileActionDeduplicator<'seen> {
+    ///
+    /// # Arguments
+    /// - `actions_batch` - A boxed [`EngineData`] instance representing a batch of actions.
+    /// - `is_log_batch` - `true` if the batch originates from a commit log, `false` if from a checkpoint.
+    ///
+    /// Returns a [`DeltaResult`] containing the processor’s output, which includes only selected actions.
+    ///
+    /// Note: Since log replay is stateful, processing may update internal processor state (e.g., deduplication sets).
+    fn process_actions_batch(
+        &mut self,
+        actions_batch: Box<dyn EngineData>,
+        is_log_batch: bool,
+    ) -> DeltaResult<Self::Output>;
+
+    /// Applies the processor to an actions iterator and filters out empty results.
+    ///
+    /// # Arguments
+    /// * `action_iter` - Iterator of action batches and their source flags
+    ///
+    /// Returns an iterator that yields the Output type of the processor.
+    fn process_batches(
+        mut self,
+        action_iter: impl Iterator<Item = DeltaResult<(Box<dyn EngineData>, bool)>>,
+    ) -> impl Iterator<Item = DeltaResult<Self::Output>> {
+        action_iter
+            .map(move |action_res| {
+                let (batch, is_log_batch) = action_res?;
+                self.process_actions_batch(batch, is_log_batch)
+            })
+            .filter(|res| {
+                // TODO: Leverage .is_none_or() when msrv = 1.82
+                res.as_ref()
+                    .map_or(true, |result| result.has_selected_rows())
+            })
+    }
+
+    /// Builds the initial selection vector for the action batch, used to filter out rows that
+    /// are not relevant to the current processor's purpose (e.g., checkpointing, scanning).
+    /// This method performs a first pass of filtering using an optional [`DataSkippingFilter`].
+    /// If no filter is provided, it assumes that all rows should be selected.
+    ///
+    /// The selection vector is further updated based on the processor's logic in the
+    /// `process_actions_batch` method.
+    ///
+    /// # Arguments
+    /// - `batch` - A reference to the batch of actions to be processed.
+    ///
+    /// # Returns
+    /// A `DeltaResult<Vec<bool>>`, where each boolean indicates if the corresponding row should be included.
+    /// If no filter is provided, all rows are selected.
+    fn build_selection_vector(&self, batch: &dyn EngineData) -> DeltaResult<Vec<bool>> {
+        match self.get_data_skipping_filter() {
+            Some(filter) => filter.apply(batch),
+            None => Ok(vec![true; batch.len()]), // If no filter is provided, select all rows
+        }
+    }
+
+    /// Returns an optional reference to the [`DataSkippingFilter`] used to filter rows
+    /// when building the initial selection vector in `build_selection_vector`.
+    /// If `None` is returned, no filter is applied, and all rows are selected.
+    fn get_data_skipping_filter(&self) -> Option<&DataSkippingFilter>;
```

Comment:
should this live here? or can it be some extension trait?

Reply:
EDIT: see other comment - let's discuss over there

Reply:
also nit: if we _do_ keep it
```suggestion
    fn data_skipping_filter(&self) -> Option<&DataSkippingFilter>;
```

Reply:
sounds like we should keep it so I'll advocate for the rename :)